Introduce PodSecurityPolicy in the policy/v1beta1 API group

php-coder opened the pull request:
Types/constants are completely the same as in `extensions/v1beta1` except that they are located outside of the `extensions` API group.

**What this PR does / why we need it**:
This is the first step for migrating PSP-related stuff away of `extensions` group. See #43214 for more information.

Also it related to https://github.com/kubernetes/features/issues/5

**Example**:
```console
$ cat restricted2.yaml 
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: restricted2
...
$ kubectl create -f restricted.yaml 
podsecuritypolicy "restricted2" created
$ kubectl get psp restricted2 -o yaml
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
...
```

**Release note**:
```release-note
The `PodSecurityPolicy` API has been moved to the `policy/v1beta1` API group. The `PodSecurityPolicy` API in the `extensions/v1beta1` API group is deprecated and will be removed in a future release. Authorizations for using pod security policy resources should change to reference the `policy` API group after upgrading to 1.11.
```

dims commented:
/ok-to-test

@php-coder you should apply for membership to the kubernetes org :)

php-coder commented on pkg/apis/psp/v1beta1/doc.go:
```diff
@@ -0,0 +1,23 @@
+*/
+
+// +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/psp
+// +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/extensions
```
@sttts I'm not sure whether I need this line or not. My theory is that it should generate conversions from/to extensions to psp types but I'm not sure. Could you tell me what it does?

sttts replied:
From Makefile.generated_files:
```
#
# Conversion generation
#
# Any package that wants conversion functions generated must include one or
# more comment-tags in any .go file, in column 0, of the form:
#     // +k8s:conversion-gen=<CONVERSION_TARGET_DIR>
#
# The CONVERSION_TARGET_DIR is a project-local path to another directory which
# should be considered when evaluating peer types for conversions.  Types which
# are found in the source package (where conversions are being generated)
# but do not have a peer in one of the target directories will not have
# conversions generated.
```

sttts replied:
I expect that you don't need this if all your types are in the new pkg/apis directory. But if you share the internal types with pkg/apis/extensions, you will need this tag.

php-coder commented on pkg/apis/psp/OWNERS:
```diff
@@ -0,0 +1,39 @@
+reviewers:
```
I copied `OWNERS` file from `pkg/apis/extensions/OWNERS`. It it OK or we should start an empty `OWNERS` file?

sttts replied:
that's a huge list. OWNERS should be more focused.

php-coder replied:
May I just remove `OWNERS` files?

sttts replied:
If it does not exist, the everything from the parent dir is inherited. But having a handful or less of reviewers here makes certainly sense.

php-coder replied:
@sttts Ok, thanks!

I decided to include only myself as a reviewer because I don't know who also want to be reviewer for this stuff.

@pweil- @simo5 @adelton @enj Let me know if you want to be reviewers and I'll add you to this list.

pweil- replied:
Sure.  Please include @liggitt @tallclair as well.

php-coder replied:
Thanks! Added.

php-coder commented:
`make update` fails with the following error:
```
Running update-generated-protobuf
Running update-codegen
Running update-generated-docs
# k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes
vendor/k8s.io/client-go/kubernetes/clientset.go:90: duplicate method PolicyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:95: duplicate method Policy
vendor/k8s.io/client-go/kubernetes/clientset.go:135: duplicate field policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:292: ambiguous selector c.policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:298: ambiguous selector c.policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:302: (*Clientset).PolicyV1beta1 redeclared in this block
	previous declaration at vendor/k8s.io/client-go/kubernetes/clientset.go:291
vendor/k8s.io/client-go/kubernetes/clientset.go:303: ambiguous selector c.policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:308: (*Clientset).Policy redeclared in this block
	previous declaration at vendor/k8s.io/client-go/kubernetes/clientset.go:297
vendor/k8s.io/client-go/kubernetes/clientset.go:309: ambiguous selector c.policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:455: ambiguous selector cs.policyV1beta1
vendor/k8s.io/client-go/kubernetes/clientset.go:455: too many errors
# k8s.io/kubernetes/vendor/k8s.io/client-go/listers/policy/v1beta1
vendor/k8s.io/client-go/listers/policy/v1beta1/eviction.go:34: undefined: EvictionListerExpansion
# k8s.io/kubernetes/pkg/client/listers/policy/internalversion
pkg/client/listers/policy/internalversion/eviction.go:34: undefined: EvictionListerExpansion
!!! [1101 12:47:43] Call tree:
!!! [1101 12:47:43]  1: /home/coder/git/src/home/kubernetes/hack/lib/golang.sh:705 kube::golang::build_binaries_for_platform(...)
!!! [1101 12:47:43]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1101 12:47:43] Call tree:
!!! [1101 12:47:43]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1101 12:47:43] Call tree:
!!! [1101 12:47:43]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [all] Error 1
```
@sttts Could you look at this PR and give me a hint what's wrong here? For some reason, generators adds generated code to the existing files and it leads to duplicated definitions. I couldn't find why they are doing this.

It also not clear where I should put the files for `policy.authorization.k8s.io`. The `policy` directory is already used, so I put the files into the `psp` directory. Is it correct?

sttts commented on pkg/apis/psp/types.go:
```diff
@@ -0,0 +1,283 @@
+/*
```
will you delete the old internal types for PSPs?

php-coder replied:
I'm following a roadmap that @soltysh gave me: in this Kubernetes release, I only need to introduce a new API that will be stored as old PSP types. So, in the current iteration, I won't remove anything from PSP.

sttts replied:
Which doubles the code for the internal types. You can easily share them by removing the old extensions types and let the extensions group point to your new ones in the new group. Compare https://github.com/kubernetes/kubernetes/pull/50327.

php-coder replied:
Maybe I'm missing something here, but I don't see how it doubles the code for the internal types. New API group is using the old internal types.

sttts commented:
> For some reason, generators adds generated code to the existing files and it leads to duplicated definitions. I couldn't find why they are doing this.
> 
> It also not clear where I should put the files for policy.authorization.k8s.io. The policy directory is already used, so I put the files into the psp directory. Is it correct?

Looks like clientgen uses the first segment of the group name for the identifiers in the clientset. @mfojtik can your exception file fix this?

enj commented:
Is `policy.authorization.k8s.io` specific enough?  RBAC could easily be thought to fit in this group (even though it does not).

liggitt commented:
yeah, I'm unsure of the group as well. what would be the unifying characteristic of objects in the `policy.authorization.k8s.io` group? For example, the `NetworkPolicy` object is in the `networking.k8s.io` group.

some possibilities:
* the `policy` group: already contains `PodDisruptionBudget` objects which are policies around pod eviction
* `podsecuritypolicy.admission.k8s.io` - objects specific to this admission plugin

php-coder commented:
> `podsecuritypolicy.admission.k8s.io` - objects specific to this admission plugin

This sounds better to me. I'll do update on next Monday to use it, unless someone object against.

sttts commented:
https://github.com/kubernetes/kubernetes/pull/54950 adds the missing support for non-unique GroupName prefixes to client-gen and informer-gen.

sttts commented:
> It also not clear where I should put the files for policy.authorization.k8s.io. The policy directory is already used, so I put the files into the psp directory. Is it correct?

There is no strict rule. Policy is for the pod policy types. Are PSPs the only types in the new group? Maybe authpolicy might make sense.

php-coder commented:
I've updated PR:
- rebased to the latest master
- `policy.authorization.k8s.io` -> `podsecuritypolicy.admission.k8s.io`

Now `make update` fails with:
```
Running update-generated-docs
# k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion/podsecuritypolicy/internalversion
pkg/client/informers/informers_generated/internalversion/podsecuritypolicy/internalversion/podsecuritypolicy.go:51: client.Podsecuritypolicy undefined (type internalclientset.Interface has no field or method Podsecuritypolicy)
pkg/client/informers/informers_generated/internalversion/podsecuritypolicy/internalversion/podsecuritypolicy.go:51: not enough arguments to return
	have (<T>)
	want (runtime.Object, error)
pkg/client/informers/informers_generated/internalversion/podsecuritypolicy/internalversion/podsecuritypolicy.go:54: client.Podsecuritypolicy undefined (type internalclientset.Interface has no field or method Podsecuritypolicy)
pkg/client/informers/informers_generated/internalversion/podsecuritypolicy/internalversion/podsecuritypolicy.go:54: not enough arguments to return
	have (<T>)
	want (watch.Interface, error)
!!! [1108 17:08:46] Call tree:
!!! [1108 17:08:46]  1: /home/coder/git/src/home/kubernetes/hack/lib/golang.sh:706 kube::golang::build_binaries_for_platform(...)
!!! [1108 17:08:46]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1108 17:08:46] Call tree:
!!! [1108 17:08:46]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1108 17:08:46] Call tree:
!!! [1108 17:08:46]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [all] Error 1
Running update-generated-docs FAILED
```
For some reason, `pkg/client/clientset_generated/internalclientset/clientset.go` wasn't updated by `update-codegen.sh` script. Any pointers would be helpful.

sttts commented:
Do we really want `podsecuritypolicy.admission.k8s.io`? I am not a fan of this group inflation where groups with one type are creates. What speaks against putting it directly into admission.k8s.io? or at least into policy.admission.k8s.io as you had before.

php-coder commented:
> What speaks against putting it directly into admission.k8s.io?

It's not clear who should make this decision. I can put it any place.

sttts commented:
Looks like `admission.k8s.io` is not served by the apiserver and it is marked as such in update-codegen.sh. So probably it is the wrong place.

Wondering, do we have a proposal somewhere which discusses where to put the PSP types?

php-coder commented:
> Looks like admission.k8s.io is not served by the apiserver and it is marked as such in update-codegen.sh. So probably it is the wrong place.

That's explain why `hack/update-swagger-spec.sh` fails to me because it tries to get `127.0.0.1:8050/swaggerapi/apis/podsecuritypolicy.admission.k8s.io/v1beta1` and found nothing.

php-coder commented:
> Wondering, do we have a proposal somewhere which discusses where to put the PSP types?

Not a proposal, but there were some comments about it: https://github.com/kubernetes/kubernetes/issues/43214#issuecomment-287040655

sttts commented:
No, I meant only admission.k8s.io. For podsecuritypolicy.admission.k8s.io you have to install the api group in the api server.

sttts commented:
@php-coder compare the registries in pkg/registry and https://github.com/sttts/kubernetes/blob/012b085ac870d359131f4251213bf2fff1d15aa0/pkg/master/master.go#L337.

php-coder commented:
> For podsecuritypolicy.admission.k8s.io you have to install the api group in the api server.

I thought that I already did it: https://github.com/kubernetes/kubernetes/pull/54933/files#diff-e5a727ab116799f561f5bfa3df80868cR229

php-coder commented:
> @php-coder compare the registries in pkg/registry and https://github.com/sttts/kubernetes/blob/012b085ac870d359131f4251213bf2fff1d15aa0/pkg/master/master.go#L337.

At this point, `podsecuritypolicy.admission.k8s.io` doesn't have it's own registry because it's supposed that it will be persisted as `extensions` API group. That's why I didn't add a registry.

As far I understand, this https://github.com/kubernetes/kubernetes/pull/54933/files#diff-36e1d7088209b1ab6ff878b18bb35210R570 should do the trick. Hm. Maybe I'm mixing storage, rest, and something else together?!..

sttts commented:
> I thought that I already did it:

That line only registers your api group in the aggregator which sits in front of the kube-apiserver. It does not implement the actual api.

sttts commented:
> At this point, podsecuritypolicy.admission.k8s.io doesn't have it's own registry because it's supposed that it will be persisted as extensions API group. That's why I didn't add a registry.

No need to re-implement the registry. Compare how this is done for the Deployment resource, living in extensions and in apps https://github.com/sttts/kubernetes/blob/26bf978c07a90d1e4f21d1c02a8de7eb01c3ad3f/pkg/registry/apps/rest/storage_apps.go#L64.

php-coder commented:
Would it be possible to include it in 1.9 or there are no chances? In other words, how many days I have for update it?

php-coder commented:
While I'm still fixing/updating the code to pass the tests, I need a feedback. I'd like to see it merged in 1.9. It's not required for our migration from SCC to PSP but would be very useful to have.

PTAL @smarterclayton @liggitt @pweil- @soltysh

mfojtik commented:
/cc @kubernetes/sig-auth-api-reviews

php-coder commented:
CC @tallclair

Also the question "is it possible to include this change in 1.9" is open.

liggitt commented on pkg/apis/podsecuritypolicy/types.go:
```diff
@@ -0,0 +1,285 @@
+
+// PodSecurityPolicy governs the ability to make requests that affect the SecurityContext
+// that will be applied to a pod and container.
+type PodSecurityPolicy struct {
```
I didn't expect a second internal type... I expected the approach we've taken with other objects as we've moved them out of the extensions group:
* define external type in destination API group
* define conversions from that external type to the existing internal type (see apps/v1/doc.go as an example)
* register external types as cohabitating resources for storage (this PR does this already)
* register existing internal type into new group's internal scheme (see apps/register.go#addKnownTypes as an example)

I would follow the approach ReplicaSet, DaemonSet, etc, took

php-coder replied:
@liggitt Thank you for the guide! I followed all the steps:

> define external type in destination API group

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/staging/src/k8s.io/api/podsecuritypolicy/v1beta1/types.go

> define conversions from that external type to the existing internal type (see apps/v1/doc.go as an example)

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/pkg/apis/podsecuritypolicy/v1beta1/doc.go

> register external types as cohabitating resources for storage

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/cmd/kube-apiserver/app/server.go#L584

> register existing internal type into new group's internal scheme

Done: https://github.com/php-coder/kubernetes/blob/psp_introduce_new_api_group/pkg/apis/podsecuritypolicy/register.go#L47-L53

But my API wasn't registered and not available:
```console
$ head -2 restricted2.yaml
apiVersion: podsecuritypolicy.admission.k8s.io/v1beta1
kind: PodSecurityPolicy
$ kubectl create -f restricted.yaml 
error: unable to recognize "restricted.yaml": no matches for podsecuritypolicy.admission.k8s.io/, Kind=PodSecurityPolicy
$ kubectl api-versions | grep podsecurity
$ curl -sk https://localhost:6443/apis | grep podsecurity
$ curl -sk https://localhost:6443/apis/podsecuritypolicy.admission.k8s.io/v1beta1
404 page not found
```
I don't understand what I'm missing here :( Could you help me?

liggitt commented:
@php-coder I don't think this is close enough for 1.9

liggitt commented:
cc @kubernetes/api-reviewers 

this is moving PSP out of the extensions API group (which is good). there's a question on what an appropriate destination group would be.

from https://github.com/kubernetes/kubernetes/pull/54933#issuecomment-341132123:

> what would be the unifying characteristic of objects [...]. For example, the `NetworkPolicy` object is in the `networking.k8s.io` group.
> 
> some possibilities:
> * the `policy` group: already contains `PodDisruptionBudget` objects which are policies around pod eviction
> * `podsecuritypolicy.admission.k8s.io` - objects specific to this admission plugin

other thoughts?

Commit:
Introduce PodSecurityPolicy in the policy/v1beta1 API group.

PSP are completely the same as in extensions/v1beta1 except that they
are located outside of the extensions API group.

php-coder commented:
The current issue is
```
I0112 18:39:20.158] +++ [0112 18:39:20] Building go targets for linux/amd64:
I0112 18:39:20.158]     ./vendor/k8s.io/code-generator/cmd/conversion-gen
W0112 18:39:21.722] make[1]: *** No rule to make target '.make/pkg/apis/podsecuritypolicy/gofiles.mk.stamp', needed by 'pkg/apis/podsecuritypolicy/v1beta1/zz_generated.conversion.go'.  Stop.
W0112 18:39:21.722] make: *** [generated_files] Error 2
I0112 18:39:21.823] Makefile:481: recipe for target 'generated_files' failed
```
I spent at least a day to debug it but no luck. I don't understand how it works, why it fails and how to fix it :( If someone told me that this cryptic message deserves an issue, I'd create one.

php-coder commented:
> make[1]: *** No rule to make target '.make/pkg/apis/podsecuritypolicy/gofiles.mk.stamp', needed by 'pkg/apis/podsecuritypolicy/v1beta1/zz_generated.conversion.go'.  Stop.

Solved by adding `pkg/apis/podsecuritypolicy/doc.go` file.

Thanks @sttts to point to `build/root/Makefile.generated_files`.

php-coder commented:
Now it fails with:
```
I0115 15:46:59.038] +++ [0115 15:46:59] Building go targets for linux/amd64:
I0115 15:46:59.038]     ./vendor/k8s.io/code-generator/cmd/conversion-gen
W0115 15:47:04.268] F0115 15:47:04.267411   20436 conversion.go:262] cannot import package ../../../../vendor/k8s.io/api/podsecuritypolicy/v1beta1
W0115 15:47:04.295] make[1]: *** [gen_conversion] Error 1
W0115 15:47:04.296] make: *** [generated_files] Error 2
I0115 15:47:04.396] Makefile.generated_files:656: recipe for target 'gen_conversion' failed
I0115 15:47:04.397] Makefile:481: recipe for target 'generated_files' failed
```

php-coder commented:
> I0115 15:47:04.396] Makefile.generated_files:656: recipe for target 'gen_conversion' failed

Solved by fixing `pkg/apis/podsecuritypolicy/v1beta1/doc.go` file:
```diff
@@ -15,7 +15,7 @@ limitations under the License.
 */
 
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/apis/podsecuritypolicy
-// +k8s:conversion-gen-external-types=../../../../vendor/k8s.io/api/podsecuritypolicy/v1beta1
+// +k8s:conversion-gen-external-types=k8s.io/api/podsecuritypolicy/v1beta1
 // +groupName=podsecuritypolicy.admission.k8s.io
 // +k8s:defaulter-gen=TypeMeta
 // +k8s:defaulter-gen-input=../../../../vendor/k8s.io/api/podsecuritypolicy/v1beta1
```

php-coder commented:
/hold cancel
/assign @smarterclayton

php-coder commented:
PTAL @kubernetes/api-approvers

tallclair commented on pkg/apis/podsecuritypolicy/doc.go:
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
nit: 2018 (same everywhere else)

php-coder replied:
Ah, yes, good catch! Fixed.

sttts commented on pkg/apis/podsecuritypolicy/register.go:
```diff
@@ -0,0 +1,53 @@
+// Adds the list of known types to api.Scheme.
+func addKnownTypes(scheme *runtime.Scheme) error {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&extensions.PodSecurityPolicy{},
```
In a follow-up we should also move the old internal extensions types here.

php-coder replied:
As far I understand, the next follow-up will be after 1.10 release. Is it correct?

sttts replied:
After this PR has merged.

php-coder replied:
It's still not clear to me:
1) why it should be in the next PR? Why we can't include it in the current one, for example?
2) what do you mean by "old internal extensions types"? Why should move them?

soltysh replied:
We've discussed this, we'll probably move the internal in the next iteration (iow. next release) along with storage and deprecation of extensions.PSP

php-coder commented:
/test pull-kubernetes-e2e-kops-aws

php-coder commented:
@liggitt It's green! PTAL.

php-coder commented:
/test pull-kubernetes-e2e-kops-aws

php-coder commented:
/test pull-kubernetes-e2e-kops-aws

php-coder commented:
@smarterclayton @liggitt PTAL

soltysh reviewed (approved):
/lgtm

soltysh commented:
@kubernetes/api-approvers for the group name + api approval

bgrant0607 commented:
@liggitt @sttts @deads2k Do we have any other examples of hierarchical API group domains? Do we have an admission.k8s.io group already? Is there an easy way to generate a list of all API groups in use in https://github.com/kubernetes/api?

liggitt commented:
> @liggitt @sttts @deads2k Do we have any other examples of hierarchical API group domains?

* `authorization.k8s.io` contains types that are not specific to any particular authorizer (SubjectAccessReview, etc)
* `rbac.authorization.k8s.io` contains all the types specific to RBAC

> Do we have an admission.k8s.io group already?

Yes - https://github.com/kubernetes/api/blob/master/admission/v1beta1/types.go

It contains types related to admission, but not to any particular admission plugin (AdmissionReview, sent/received from admission webhooks)

> Is there an easy way to generate a list of all API groups in use in https://github.com/kubernetes/api?

I'm not sure... I think the current generation happens from the discovery docs published by the API server, which would exclude groups that are defined but not served from kube-apiserver (config types (#50041), types for webhook callouts, etc). It would be good to be able to generate openapi schema definitions and docs from the declared types the same way we do with clients/informers, etc, without requiring starting a server serving the types. cc @mbohlool

liggitt commented:
guidance from the [architecture call](https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit):

* avoid group names that pin to single resource
* avoid group names that pin to a particular mechanism (if a way other than admission became available to apply PodSecurityPolicy, we wouldn't want "admission" in the API group name)
* focus on area in the group name (policy, security policy, etc)

we'll come up with a couple ideas that follow that guidance and regroup (pun totally intended)

bgrant0607 commented:
@liggitt I was thinking about the policy.k8s.io group. I could imagine moving ResourceQuota and LimitRange into that group eventually. I think PodSecurityPolicy could fit, also. 

Layer/group diagram:
https://docs.google.com/presentation/d/1P7lg13Rw21NQ59ts22PTzI1q81nnYVK-cOVhYTzg9tg/edit#slide=id.g285efa6b8e_0_220

I'll ask about the PodDisruptionBudget roadmap.

php-coder commented:
> I was thinking about the policy.k8s.io group.

I don't mind and can update the PR if @kubernetes/api-approvers is OK with that group. Any objections?

php-coder commented:
@kubernetes/api-approvers Ping

php-coder commented:
@bgrant0607 Any updates on that?

php-coder commented:
> I was thinking about the policy.k8s.io group.

@bgrant0607 There is no such group as `policy.k8s.io` but there is `policy` API group. Here is [the comment](https://github.com/kubernetes/kubernetes/blob/c17b418f89e056bea272ed46ca9f63f8b1e37589/pkg/apis/policy/v1beta1/doc.go#L22-L24) for this group:

> Package policy is for any kind of policy object.  Suitable examples, even if they aren't all here, are policyv1beta1.PodDisruptionBudget, PodSecurityPolicy, NetworkPolicy, etc.

I'm going to update this PR and move PSP into `policy`.

php-coder commented:
> policy.k8s.io works for me (I'd prefer we make the group name match the chosen k8s.io suffixed convention)

Thanks! I'll create a new group for it.

I have a question: where to put files then? `pkg/apis/policy/` and `staging/src/k8s.io/api/policy/` dirs are used for `policy` group. AFIAU I can't put files for `policy.k8s.io` there.

Commit:
Update examples to use PSPs from the policy API group.

tallclair commented:
I'm supportive of bundling this with other policies (I'd like to see more convergence of the policy mechanisms in general).

We shouldn't have `policy` and `policy.k8s.io` though. What are the steps for migrating `policy` to `policy.k8s.io`? Unless that move happens in v1.10, I think this should just go into `policy` now, and then move the whole group to `policy.k8s.io` together in a later release.

php-coder commented:
> Unless that move happens in v1.10, I think this should just go into policy now

Would we need to bump `policy` version to `v1beta2` in this case?

liggitt commented:
> What are the steps for migrating policy to policy.k8s.io? 

 The same as moving from the extensions group to the policy group :-(

Define new types, define cross group conversions, maintain the old group for the deprecation period, and indefinitely until storage migration is solved

tallclair commented:
> The same as moving from the extensions group to the policy group :-(

But the code location and all internal code would stay the same... right? Could we solve this by either adding a `v1beta2` with the groupname, or simply rename v1beta1, and make `/apis/policy/v1beta1` an alias to `/apis/policy.k8s.io/v1beta1`?

php-coder commented:
@bgrant0607 Ping. Your approval for continuing with `policy.k8s.io` is required.

/hold

bgrant0607 commented:
@php-coder Sorry, but github notifications are till next to useless for me.

What is the proposal? To migrate PodSecurityPolicy and PodDisruptionBudget to policy.k8s.io?

I'm fine with that.

liggitt commented:
> @tallclair:
> EDIT: to clarify, I mean moving the types that are already in the policy group to a policy.k8s.io group.

I spent a while experimenting with migrating `policy/v1beta1` to `policy.k8s.io/v1beta2` as a first step, and decided it wasn't worth the fight with our generators, machinery, and client interfaces:
* None of the generators were happy with sharing the internal `policy` folder between two groups
* The client interface structure made it impossible to have `clientset.Policy().V1beta1()` go to `policy/v1beta1` and `clientset.Policy().V1beta2()` go to `policy.k8s.io/v1beta2`

Not only that, but I forgot that `policy/v1beta1` eviction resources are served under `pods/eviction`, so we already are serving API objects from the `policy` group from the v1 API. I think we're stuck with the existing group name.

Let's just add PodSecurityPolicy to `policy/v1beta1` and move on.

Commit:
Update generated files.

php-coder commented:
> Let's just add PodSecurityPolicy to policy/v1beta1 and move on.

PR has been updated for that. PTAL.

/hold cancel

liggitt commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -598,6 +599,7 @@ func BuildStorageFactory(s *options.ServerRunOptions, apiResourceConfig *servers
 	storageFactory.AddCohabitatingResources(apps.Resource("daemonsets"), extensions.Resource("daemonsets"))
 	storageFactory.AddCohabitatingResources(apps.Resource("replicasets"), extensions.Resource("replicasets"))
 	storageFactory.AddCohabitatingResources(api.Resource("events"), events.Resource("events"))
+	storageFactory.AddCohabitatingResources(extensions.Resource("podsecuritypolicies"), policy.Resource("podsecuritypolicies"))
```
add a TODO to flip the order here in 1.11

liggitt commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -296,6 +296,11 @@ var etcdStorageData = map[schema.GroupVersionResource]struct {
 	},
+	gvr("policy", "v1beta1", "podsecuritypolicies"): {
+		stub:             `{"metadata": {"name": "psp2"}, "spec": {"fsGroup": {"rule": "RunAsAny"}, "privileged": true, "runAsUser": {"rule": "RunAsAny"}, "seLinux": {"rule": "MustRunAs"}, "supplementalGroups": {"rule": "RunAsAny"}}}`,
+		expectedEtcdPath: "/registry/podsecuritypolicy/psp2",
```
in SpecialDefaultResourcePrefixes, add `{Group: "policy", Resource: "podsecuritypolicies"}: "podsecuritypolicy",` to make sure the current prefix remains the same when we flip the preferred version in 1.11

tallclair commented on pkg/apis/policy/register.go:
```diff
@@ -48,6 +49,8 @@ func addKnownTypes(scheme *runtime.Scheme) error {
 		&PodDisruptionBudget{},
 		&PodDisruptionBudgetList{},
+		&extensions.PodSecurityPolicy{},
+		&extensions.PodSecurityPolicyList{},
```
Did you mean to register the extensions one here?

liggitt replied:
This is internal, which hubs to the types located in extensions (same as other types that are moving between API groups)

tallclair commented on pkg/registry/policy/rest/storage_policy.go:
```diff
@@ -24,6 +24,7 @@ import (
 	serverstorage "k8s.io/apiserver/pkg/server/storage"
 	"k8s.io/kubernetes/pkg/api/legacyscheme"
 	"k8s.io/kubernetes/pkg/apis/policy"
+	pspstore "k8s.io/kubernetes/pkg/registry/extensions/podsecuritypolicy/storage"
```
Should this be extensions or policy?

liggitt replied:
There's only a single registry and it lives under extensions for now. We'll move it as a follow up, but it's internal only

tallclair commented on examples/podsecuritypolicy/rbac/policies.yaml:
```diff
@@ -23,7 +23,7 @@ spec:
     max: 65536
 ---
-apiVersion: extensions/v1beta1
+apiVersion: policy/v1beta1
```
I was going to suggest updating the addons too, but then I realized that could result in duplication of all of the policies after an upgrade...

liggitt replied:
> was going to suggest updating the addons too, but then I realized that could result in duplication of all of the policies after an upgrade...

it shouldn't... the old and new API group are colocated... existing extensions/v1beta1 objects would also surface under policy/v1beta1

tallclair replied:
oh, great. In that case, @php-coder can you update the addons too?

tallclair commented:
Maybe update the e2e tests too?

liggitt commented:
> Maybe update the e2e tests too? (or it could be a separate PR - same with the addons)

Yeah, typically we wait a release to ensure addons, controllers, and e2e tests continue to work across a HA master.

There's a long follow-up set of work to do for this move:
* switch admission plugin to use policy group
* switch authorization check for `use` verb (will need to check both groups for some time period)
* update addon manifests (in a future release)
* update e2e tests (test both for some time period)
* move internal types to policy package
* move registry to policy package
* switch preferred storage version to policy group (in a future release)

tallclair commented:
Thanks for the clarification. SGTM.

php-coder commented:
The current status: `TestRoundTripTypes` fails with the error:
```
roundtrip.go:286: PodSecurityPolicy: converting (extensions.PodSecurityPolicy).RunAsUser.Ranges.[0] to (v1beta1.PodSecurityPolicy).RunAsUser.Ranges.[0]: type names don't match (UserIDRange, IDRange), and no conversion 'func (extensions.UserIDRange, v1beta1.IDRange) error' registered.
```

liggitt commented on pkg/apis/policy/v1beta1/defaults.go:
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
nit, 2018 in all new files

php-coder replied:
There was only one single file.

liggitt commented:
one nit on dates, lgtm otherwise

cc @kubernetes/sig-auth-api-reviews @kubernetes/api-approvers

php-coder commented:
@liggitt The comment was addressed. Thank you for guidance and patience!

liggitt commented:
/lgtm
/hold for ack from @kubernetes/api-approvers

bgrant0607 commented:
@liggitt ACK that `policy` is fine. Thanks much

liggitt commented:
/hold cancel

php-coder commented:
@liggitt Are we waiting for someone else LGTM from API approvers or I can ask examples/ owners to review their part?

smarterclayton commented:
/approve 

examples
